lib/ruby_lsp/requests/path_completion.rb

```diff
@@ -0,0 +1,99 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # ![Path completion demo](../../misc/path_completion.gif)
+    #
+    # The [completion](https://microsoft.github.io/language-server-protocol/specification#textDocument_completion)
+    # request looks up Ruby files in the $LOAD_PATH to suggest path completion inside `require` statements.
+    #
+    # # Example
+    #
+    # ```ruby
+    # require "ruby_lsp/requests" # --> completion: suggests `base_request`, `code_actions`, ...
+    # ```
+
+    class PathCompletion < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, position: Document::PositionShape).void }
+      def initialize(document, position)
+        super(document)
+
+        @tree = T.let(Support::PrefixTree.new(collect_load_path_files), Support::PrefixTree)
+
+        char_position = document.create_scanner.find_char_position(position)
+        @target = T.let(find(T.must(document.tree), char_position), T.nilable(SyntaxTree::TStringContent))
+      end
+
+      sig { override.returns(T.all(T::Array[LanguageServer::Protocol::Interface::CompletionItem], Object)) }
+      def run
+        # # no @target means the we are not inside a `require` call
+        return [] unless @target
+
+        text = @target.value
+        @tree.search(text).map do |path|
+          build_completion(path, path.delete_prefix(text))
+        end
+      end
+
+      private
+
+      sig { returns(T::Array[String]) }
+      def collect_load_path_files
+        $LOAD_PATH.flat_map do |p|
+          Dir.glob("**/*.rb", base: p)
+        end.map! do |result|
+          result.delete_suffix!(".rb")
+        end.sort
+      end
+
+      sig do
+        params(
+          node: SyntaxTree::Node,
+          position: Integer,
+        ).returns(T.nilable(SyntaxTree::TStringContent))
+      end
+      def find(node, position)
+        matched =
+          node.child_nodes.compact.bsearch do |child|
+            if (child.location.start_char..child.location.end_char).cover?(position)
+              0
+            else
+              position <=> child.location.start_char
+            end
+          end
+
+        case matched
+        when SyntaxTree::Command, SyntaxTree::CommandCall, SyntaxTree::CallNode
+          return unless matched.message.value == "require"
+
+          args = matched.arguments
+          if args.is_a?(SyntaxTree::ArgParen)
+            args = args.arguments
+          end
+
+          path_node = args.parts.first.parts.first
+          return unless path_node
+          return unless (path_node.location.start_char..path_node.location.end_char).cover?(position)
```

Comment:
Didn't we already check this when computing `matched`?

Reply:
Maybe we should extract this as a method so we can attach it to a more explicit name